### PR TITLE
Rewrite notifications system, hide RxJava even more deeper. Fix flaky test with concurrent notification

### LINKED ISSUE #326: Unstable test NotifyAboutChangesTest#notifyAboutChangesConcurrently()
Probably it's problem of `rx.observers.TestObserver`, as I see it uses non thread safe `ArrayList` as storage for notifications and received items.

https://github.com/ReactiveX/RxJava/issues/2947

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Closes #326 

Thanks to guys from RxJava, I found a problem in the notifications code.

@nikitin-da PTAL

### CODE REVIEW

**Comment by nikitin-da** [comment]
retest this please

**Comment by nikitin-da** [comment]
It seems like it doesn't receive notifications again =(

**Comment by artem-zinnatullin** [comment]
I'll take a look tonight

**Comment by artem-zinnatullin** [comment]
retest this please

**Comment by artem-zinnatullin** [comment]
Don't merge please

**Comment by nikitin-da** [comment]
Ok

File: storio-common/src/main/java/com/pushtorefresh/storio/internal/ChangesBus.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2015 Artem Zinnatullin.
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/internal/ChangesBus.java]: Licence should be removed, incorrect Android Studio settings..

File: storio-common/src/main/java/com/pushtorefresh/storio/internal/RxChangesBus.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2015 Artem Zinnatullin.
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/internal/RxChangesBus.java]: Remove licence

File: storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/NotifyAboutChangesTest.java [hunk 3]
```diff
@@ -48,9 +49,9 @@ public void notifyAboutChange() {
 
     @Test
     public void notifyAboutChangesConcurrently() {
-        final int numberOfThreads = 10;
+        final int numberOfThreads = 100;
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/NotifyAboutChangesTest.java]: @nikitin-da do you feel concurrency?)

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/impl/DefaultStorIOSQLite.java [hunk 4]
```diff
@@ -45,14 +43,7 @@
...
 
-    /**
-     * Reactive bus for notifying observers about changes in StorIOSQLite
-     * One change can affect several tables, so we use {@link Changes} as representation of changes
-     */
-    @Nullable
-    private final PublishSubject<Changes> changesBus = Environment.IS_RX_JAVA_AVAILABLE
-            ? PublishSubject.<Changes>create()
-            : null;
+    private final ChangesBus<Changes> changesBus = new ChangesBus<Changes>();
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/impl/DefaultStorIOSQLite.java]: Add `@NonNull`

File: storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/NotifyAboutChangesTest.java (hunk 3 above)
Comment (nikitin-da) [inline storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/NotifyAboutChangesTest.java]: A little bite)

**Comment by artem-zinnatullin** [comment]
retest this please

**Comment by artem-zinnatullin** [comment]
(I am playing with JUnit Plugin for Jenkins)

File: storio-common/src/main/java/com/pushtorefresh/storio/internal/ChangesBus.java (hunk 1 above)
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/internal/ChangesBus.java]: Fixed

File: storio-common/src/main/java/com/pushtorefresh/storio/internal/RxChangesBus.java (hunk 2 above)
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/internal/RxChangesBus.java]: Fixed

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/impl/DefaultStorIOSQLite.java (hunk 4 above)
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/impl/DefaultStorIOSQLite.java]: Fixed

**Comment by artem-zinnatullin** [comment]
Rebased and squashed into one commit, @nikitin-da PTAL

**Comment by artem-zinnatullin** [comment]
I think I know why your emission tests sometimes fail — thread-safety.

File: storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/BaseSubscriptionTest.java [hunk 5]
```diff
@@ -14,11 +15,15 @@
...
+
+        @NonNull
         private final Queue<T> expected;
+
+        @NonNull
         private CountDownLatch lock;
 
         public BaseEmissionChecker(@NonNull Queue<T> expected) {
-            this.expected = expected;
+            this.expected = new ConcurrentLinkedQueue<T>(expected);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/BaseSubscriptionTest.java]: This queue can be accessed from multiple threads!

**Comment by artem-zinnatullin** [comment]
Done!

File: storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/BaseSubscriptionTest.java (hunk 5 above)
Comment (nikitin-da) [inline storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/BaseSubscriptionTest.java]: Damn

Comment (nikitin-da) [inline storio-sqlite/src/androidTest/java/com/pushtorefresh/storio/sqlite/impl/BaseSubscriptionTest.java]: Great fix, thank you very much

**Pull request merged**